pyccel/codegen/printing/cwrappercode.py

```diff
@@ -159,7 +198,170 @@ def get_cast_function_call(self, cast_type, arg):
+        """
+        Responsible for collecting value and managing error and create the body
+        of optional arguments in format
+                if (pyobject == Py_None){
+                    collect Null
+                }else if(Type Check == False){
+                    Print TypeError Wrong Type
+                    return Null
+                }else{
+                    assigne pyobject value to tmp variable
+                    collect the adress of the tmp variable
+                }
+        Parameters:
+        ----------
+        tmp_variable : Variable
+            The temporary variable  to hold result
+        Variable : Variable
+            The optional variable
+        collect_var : variable
+            the pyobject type variable  holder of value
+        check_type : Boolean
+            True if the type is needed
+
+        Returns
+        -------
+        body : list
+            A list of statements
+        """
+        body = [(PyccelEq(VariableAddress(collect_var), VariableAddress(Py_None)),
+                [Assign(VariableAddress(variable), variable.value)])]
+        if check_type : # Type check
+            check = PyccelNot(PythonType_Check(variable, collect_var))
+            error = PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(variable, variable.dtype))
+            body += [(check, [error, Return([Nil()])])]
+        body += [(LiteralTrue(), [Assign(tmp_variable, self.get_collect_function_call(variable, collect_var)),
+                    Assign(VariableAddress(variable), VariableAddress(tmp_variable))])]
+        body = [If(*body)]
+
+        return body
+
+    def _body_valued_variable(self, variable, collect_var, check_type = False) :
+        """
+        Responsible for collecting value and managing error and create the body
+        of valued arguments in format
+                if (pyobject == Py_None){
+                    collect default value
+                }else if(Type Check == False){
+                    Print TypeError Wrong Type
+                    return Null
+                }else{
+                    collect the value from PyObject
+                }
+        Parameters:
+        ----------
+        Variable : Variable
+            The optional variable
+        collect_var : variable
+            the pyobject type variable  holder of value
+        check_type : Boolean
+            True if the type is needed
+
+        Returns
+        -------
+        body : list
+            A list of statements
+        """
+        body = [(PyccelEq(VariableAddress(collect_var), VariableAddress(Py_None)),
+                [Assign(variable, variable.value)])]
+        if check_type : # Type check
+            check = PyccelNot(PythonType_Check(variable, collect_var))
+            error = PyErr_SetString('PyExc_TypeError', '"{} must be {}"'.format(variable, variable.dtype))
+            body += [(check, [error, Return([Nil()])])]
```

Comment:
This seems to be duplicated in a few places right?

Reply:
Maybe we can create a function `check_for_error(condition, err_msg)`?

Reply:
this can be a good idea but this function will be working only in the functions that create body.

Reply:
after all, i don't think if it is good to create the function the code will be less comprehensible and I prefer that we leave the check and the error creation separated for the moment as we don't know what kind of error we will encounter.

Reply:
I meant a function that returns an If not a tuple, but you might be right, and that doesn't save many (any?) lines anyway